File: 560-subarray-sum-equals-k/560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int res = 0;
        
        for(int i = 0, sum = 0; i < nums.length; i++) {
            sum += nums[i];
            int dif = sum - k;
            
            if (dif == 0) res++;
            res += map.getOrDefault(dif, 0);
            
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
        
        return res;
    }
}

File: 560-subarray-sum-equals-k/560-subarray-sum-equals-k.scala
import scala.collection.mutable._

object Solution {
    def subarraySum(nums: Array[Int], k: Int): Int = {
        val map: Map[Int, Int] = Map()
        var sum = 0
        var res = 0
        
        nums.foreach { n =>
            sum += n
            val diff = sum - k
            res += map.get(diff).getOrElse(0)
            if (sum == k) res += 1
            map += (sum -> (map.get(sum).getOrElse(0) + 1))
        }

        res
    }
}
